airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/factory.py

```diff
@@ -115,20 +118,43 @@ def create_component(self, component_definition: ComponentDefinition, config: Co
             class_name = CLASS_TYPES_REGISTRY[kwargs.pop("type")]
         else:
             raise ValueError(f"Failed to create component because it has no class_name or type. Definition: {component_definition}")
-        return self.build(class_name, config, **kwargs)
+        kwargs.pop("config", None)
+        return self.build(
+            class_name,
+            config,
+            instantiate,
+            **kwargs,
+        )
 
-    def build(self, class_or_class_name: Union[str, Type], config, **kwargs):
+    def build(self, class_or_class_name: Union[str, Type], config, instantiate: bool = True, **kwargs):
         if isinstance(class_or_class_name, str):
             class_ = self._get_class_from_fully_qualified_class_name(class_or_class_name)
         else:
             class_ = class_or_class_name
 
         # create components in options before propagating them
         if OPTIONS_STR in kwargs:
-            kwargs[OPTIONS_STR] = {k: self._create_subcomponent(k, v, kwargs, config, class_) for k, v in kwargs[OPTIONS_STR].items()}
+            kwargs[OPTIONS_STR] = {
+                k: self._create_subcomponent(k, v, kwargs, config, class_, instantiate) for k, v in kwargs[OPTIONS_STR].items()
+            }
 
-        updated_kwargs = {k: self._create_subcomponent(k, v, kwargs, config, class_) for k, v in kwargs.items()}
-        return create(class_, config=config, **updated_kwargs)
+        updated_kwargs = {k: self._create_subcomponent(k, v, kwargs, config, class_, instantiate) for k, v in kwargs.items()}
+
+        if instantiate:
+            return create(class_, config=config, **updated_kwargs)
+        else:
+            # Because the component's data fields definitions use interfaces, we need to resolve the underlying types into the
+            # concrete classes that implement the interface before generating the schema
+            DeclarativeComponentFactory._transform_interface_to_union(class_)
```

Comment:
Interestingly, even though we make a copy of the class and replace those annotations, it does seem to modifying the incoming class argument. I tried returning the relevant class and the way it is currently shown, to the same effect of expansion. However, if I comment this line out, the expansion will not go through (which is expected)

Reply:
this is almost certainly because the fields don't get copied.

Looking at `dataclasses.fields`, it seems like they are in the "__dataclass_fields__" field of the class' `__dict__`.

They don't get copied because we don't do a deepcopy of the dict [here](https://github.com/airbytehq/airbyte/pull/15543/files#diff-4795f61b18fa2fe8c208da8b2c2aa23d5b6674c8e2646588c819c3696839d117R274).

That being said, I don't know if there's an easy way to copy them because `mappingproxy`s can't be serialized.

I doubt this matters much...